name-registry/src/main.sw

```diff
@@ -0,0 +1,154 @@
+contract;
+
+dep data_structures;
+dep errors;
+dep events;
+dep interface;
+
+use data_structures::Record;
+use errors::{AssetErrors, AuthorisationErrors, ValidityErrors};
+use events::{
+    IdentityChangedEvent,
+    NameRegisteredEvent,
+    OwnerChangedEvent,
+    RegistrationExtendedEvent,
+};
+use interface::NameRegistry;
+use std::{
+    block::timestamp,
+    chain::auth::msg_sender,
+    constants::BASE_ASSET_ID,
+    context::call_frames::msg_asset_id,
+    context::msg_amount,
+    logging::log,
+    storage::StorageMap,
+};
+
+storage {
+    /// A mapping of names to an option of records, with a none representing an unregistered name
+    names: StorageMap<str[8], Option<Record>> = StorageMap {},
+}
+
+// TODO: Change the static 8 length str with a dynamic string when possible
+impl NameRegistry for Contract {
+    #[storage(read, write)]
+    fn extend(name: str[8], duration: u64) {
+        require(storage.names.get(name).is_some(), ValidityErrors::NameNotRegistered);
+        require((duration / 100) * price_per_hundred <= msg_amount(), AssetErrors::InsufficientPayment);
+        require(msg_asset_id() == BASE_ASSET_ID, AssetErrors::IncorrectAssetSent);
```

Comment:
Ah, I see. I suggest taking the base asset and putting it inside the config as a config time constant to allow a user to configure their registry without having to change the code.

In the future, this could be expanded to use a registry of available assets and the user gets to select which asset they want to pay in and as long as the asset is specified in that registry and the calculation for the price is met then they can use any asset they like from the curated list.

Reply:
I would also switch the order of the price and base asset checks for semantic reasons.

Do you go to a store and first count the amount and then check that the correct currency is used or do you first notice the currency and then tally up?

Just to drive the point home, you wouldn't count 100 carrots and then at the end say "sorry, these are carrots", you'd automatically look at me like I'm crazy and reject the carrots without counting them.

Reply:
I'm wondering if we should restrict this to only allow the owner to extend the name. I'm indifferent either way but I'd like to hear your thoughts.

Reply:
People may want to use different addresses to extend their names, so I would prefer not having the owner extend it as a requirement